src/chat/tools/listTopics.ts

```diff
@@ -0,0 +1,170 @@
+    }
+
+    let environmentId = params.environmentId;
+    if (!environmentId) {
+      return new LanguageModelToolResult([
+        new LanguageModelTextPart("No environment ID provided."),
+      ]);
+    }
+
+    let kafkaClusterId = params.kafkaClusterId;
+    if (!kafkaClusterId) {
+      return new LanguageModelToolResult([
+        new LanguageModelTextPart("No Kafka cluster ID provided."),
+      ]);
+    }
+
+    // Handle cases where all IDs are the same (local setup)
+    if (kafkaClusterId === environmentId) {
+      return new LanguageModelToolResult([
+        new LanguageModelTextPart(
+          `Kafka cluster ID and environment ID are the same (${kafkaClusterId}). Please use get_environments to retrieve the Kafka cluster ID.`,
+        ),
+      ]);
+    }
+
+    const loader = ResourceLoader.getInstance(connectionId);
+
+    const kafkaClusters: KafkaCluster[] =
+      await loader.getKafkaClustersForEnvironmentId(environmentId);
+
+    if (!(Array.isArray(kafkaClusters) && kafkaClusters.length)) {
+      return new LanguageModelToolResult([
+        new LanguageModelTextPart("No Kafka clusters found for the given environment ID."),
+      ]);
+    }
+
+    const kafkaCluster = kafkaClusters.find((cluster) => cluster.id === kafkaClusterId);
+    if (!kafkaCluster) {
+      return new LanguageModelToolResult([
+        new LanguageModelTextPart(
+          `No Kafka cluster found for the given ID (${kafkaClusterId}). Available clusters: ${kafkaClusters
+            .map((c) => c.id)
+            .join(", ")}`,
+        ),
+      ]);
+    }
+
+    const topics = await loader.getTopicsForCluster(kafkaCluster);
+    if (!(Array.isArray(topics) && topics.length)) {
+      logger.debug(`No topics found for cluster ${kafkaClusterId}`);
+      return new LanguageModelToolResult([
+        new LanguageModelTextPart("No topics found for the given Kafka cluster."),
+      ]);
+    }
+
+    let sampleTopics: KafkaTopic[] = [...topics]; // Create a copy of the topics array
+    const topicNameSubstring = params.topicNameSubstring;
+
+    if (topicNameSubstring) {
+      sampleTopics = sampleTopics.filter((topic) =>
+        topic.name.toLowerCase().includes(topicNameSubstring.toLowerCase()),
+      );
+      logger.debug(
+        `Filtered topics by substring "${topicNameSubstring}": ${sampleTopics.length} matches`,
+      );
+    }
+
+    // if sample topics is more than 30, slice it to 30
+    if (sampleTopics.length > 30) {
+      sampleTopics = sampleTopics.slice(0, 30);
+      logger.debug(`Limited topics to 30 samples`);
+    }
+
+    if (token.isCancellationRequested) {
+      logger.debug("Tool invocation cancelled");
+      return new LanguageModelToolResult([]);
+    }
+
+    logger.debug(`Summarizing ${sampleTopics.length} topics`);
+    return new LanguageModelToolResult([new LanguageModelTextPart(summarizeTopics(sampleTopics))]);
```

Comment:
One problem with making `summarizeTopics` handle an array is it's creating a single string (and single `LanguageModelTextPart`) no matter how many are passed, which will then impact `processInvocation`'s `result.content.length`:
<img width="882" alt="image" src="https://github.com/user-attachments/assets/31e7cc71-9df9-4be5-b17f-60fb53fa0527" />

What do you think about changing it so the summarizer is only responsible for a single `KafkaTopic` and each entry is passed into a single `LanguageModelTextPart`?